## `Table of Contents`
1. [Introduction](#section1)<br>
2. [Problem Statment](#section2)<br>
3. [Data Loading and Description](#section3)<br>
4. [Data Profiling](#section4)
    - 4.1 [Understanding the Dataset](#section401)<br/>
    - 4.2 [Pre Profiling](#section402)<br/>
    - 4.3 [Preprocessing](#section403)<br/>
    - 4.4 [Post Profiling](#section404)<br/>


<a id=section1></a> 
### 1. Introduction 

The data was collected by a Ukrainian person in 2016. Its source is the Car sales advertisement. It consists of car features like Makers, Engine type, Body Type mileage, and price, etc. These cars are Used cars, not a brand new car So it opens the possibility to analyses different aspects of car sales opportunity.

<a id=section2></a>
### 2. Problem Statment




    
The sole purpose of this notebook is to do __Exploratory Data Analysis(EDA):__
<font color=blue|yellow|>
 - What are all different kinds of cars is available in the Ukrainian car fleet?
 - To Stablished correlation among different car features.
 - To Identify factors on which price and Mileage are dependent.
 - To do a relative study of similar features among the different brands.

To achieve the above milestone I will do __Data Wrangling__.
</font>


<a id =section3></a>
### 3. Data Loading and Description
__i.__ Data Source 
- Data is collected from the Ukrainian car fleet advertisements  by a person in 2016. 

__ii.__ Dataset Brief

- The dataset consists of the information about car featurs. Its comprises of __`9576 rows of 10 columns.`__  Below is a table showing names of all the columns and their description.


| Column Name   |  Description                                               |
| :------------ |:-------------                                             | 
| Car   		| Manufacturer brand                                        | 
| Price      	| Seller’s price in advertisement (in USD)                  |  
| Body	        | Car body type                                             | 
| Mileage       | As Per advertisement(in Km)                               |   
| EngV          | Rounded engine volume (‘000 cubic cm)                     |
| EngType       | Type of fuel (“Other” in this case should be treated as NA)       |
| Registration  | Whether car registered in Ukraine or not                  |
| Year          | year of production                                        |
| Model         | Model name                                                |
| Drive         | Drive type                                                |



<font color=blue|yellow|>
    
__Disclaimer from Data collector :__ Data has gaps, so be careful and check for NA’s. I tried to check and drop repeated offers, but theoretically duplications are possible.
 </font>


#### Importing packages

In [1]:
import numpy as np                                      # It Help in process multi-dimensional arrays and matrices. It also provide  a huge collection of high-level mathematical functions
import pandas as pd                                     # It is used for data analysis. It offers data structures and operations for manipulating numerical tables and time series.
import pandas_profiling as profiling                                 # Pandas profiling provides analysis like type, unique values, missing values, quantile statistics, mean, mode, median, standard deviation, sum, skewness, frequent values, histograms, correlation between variables, count, heatmap visualization
import matplotlib.pyplot as plt                         # It provide api to visualize data. Matplotlib consists of several plots like line, bar, scatter, histogram etc
import seaborn as sns                                   # It is also data visualization library based on matplotlib. It provides a high-level interface for drawing attractive and informative statistical graphics.
%matplotlib inline
sns.set()
DATA_PATH ="https://raw.githubusercontent.com/insaid2018/Term-1/master/Data/Projects/car_sales.csv"  # Path for data set. It is github path

#### Importing the Dataset

In [2]:
carFeatureDf = pd.read_csv(DATA_PATH,encoding = "ISO-8859-1")     # encodind =UTF-8 was throwing error during parsing. Beacuse It parse 2char together . 

<a id =section4></a>
#### 4. Data Profiling
- In this section, We will __understand rows, columns and verities__ of the dataset with the help of pandas functions.
- Below analysis will be perform to understand dataset
    - Get Initial rows
	- Get Last rows
	- Get total rows and columns
	- Columns Data Type
	- Meta-info of data set, like columns dataType, missing values 
	- Brief infer of dataset
    - Null values count in Dataset
- With the help of __pandas profiling__, I will do initial Statistical analysis to get a sense of data spread, important columns, missing values, and correlations
- The __preprocessing__ section is dedicated to Data transformation like data labelling, replacing null or NA values or modifies the column values if required.
- In __Post profiling__ , I will do again Pandas profilling and validate the data again 

<a id="section401"></a>
### 4.1 Understanding the Dataset

In [3]:
carFeatureDf.head(10)          # It returns 1st 10 rows

car    price       body  mileage  engV engType registration  \
0           Ford  15500.0  crossover       68   2.5     Gas          yes   
1  Mercedes-Benz  20500.0      sedan      173   1.8     Gas          yes   
2  Mercedes-Benz  35000.0      other      135   5.5  Petrol          yes   
3  Mercedes-Benz  17800.0        van      162   1.8  Diesel          yes   
4  Mercedes-Benz  33000.0      vagon       91   NaN   Other          yes   
5         Nissan  16600.0  crossover       83   2.0  Petrol          yes   
6          Honda   6500.0      sedan      199   2.0  Petrol          yes   
7        Renault  10500.0      vagon      185   1.5  Diesel          yes   
8  Mercedes-Benz  21500.0      sedan      146   1.8     Gas          yes   
9  Mercedes-Benz  22700.0      sedan      125   2.2  Diesel          yes   

   year    model  drive  
0  2010     Kuga   full  
1  2011  E-Class   rear  
2  2008   CL 550   rear  
3  2012    B 180  front  
4  2013  E-Class    NaN  
5  2013  X-Trail   full  
6  2003   Accord  front  
7  2011   Megane  front  
8  2012  E-Class   rear  
9  2010  E-Class   rear

In [4]:
carFeatureDf.tail(10)           #It returns last 10 rows

car    price       body  mileage  engV engType registration  \
9566            UAZ    850.0        van      255   NaN   Other          yes   
9567           Jeep  13990.0  crossover      210   3.0  Diesel          yes   
9568  Mercedes-Benz   5000.0      vagon      248   2.0     Gas          yes   
9569     Volkswagen  18000.0  crossover      155   2.0  Diesel          yes   
9570         Toyota  14474.0  crossover      115   2.0  Petrol          yes   
9571        Hyundai  14500.0  crossover      140   2.0     Gas          yes   
9572     Volkswagen   2200.0      vagon      150   1.6  Petrol          yes   
9573  Mercedes-Benz  18500.0  crossover      180   3.5  Petrol          yes   
9574          Lexus  16999.0      sedan      150   3.5     Gas          yes   
9575           Audi  22500.0      other       71   3.6  Petrol          yes   

      year           model  drive  
9566  1981            3962    NaN  
9567  2006  Grand Cherokee   full  
9568  1996         E-Class   rear  
9569  2011          Tiguan   full  
9570  2007           Rav 4   full  
9571  2011          Tucson  front  
9572  1986       Passat B2  front  
9573  2008          ML 350   full  
9574  2008          ES 350  front  
9575  2007              Q7   full

In [5]:
carFeatureDf.shape

(9576, 10)

<font color=blue>
    
__It is having `9576 no. of rows` and `10 columns`__

</font>

In [6]:
carFeatureDf.columns

Index(['car', 'price', 'body', 'mileage', 'engV', 'engType', 'registration',
       'year', 'model', 'drive'],
      dtype='object')

<font color=blue>
    
__It is having `10 columns`__

</font>
<span style='color:green'>
car<br>
price<br>
body <br>
mileage<br>
engV<br>
engType <br>
registration<br>
year<br>
model<br>
drive
</span>

In [7]:
carFeatureDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9576 entries, 0 to 9575
Data columns (total 10 columns):
car             9576 non-null object
price           9576 non-null float64
body            9576 non-null object
mileage         9576 non-null int64
engV            9142 non-null float64
engType         9576 non-null object
registration    9576 non-null object
year            9576 non-null int64
model           9576 non-null object
drive           9065 non-null object
dtypes: float64(2), int64(2), object(6)
memory usage: 748.2+ KB


<span style='color:blue'>
    
- __engV__ and __drive__ column is having missing data.
- __engv__ and __price__ is having decimal values (float Data type).
- __Year__ is of int type.
- Remaining columns is of Object type. So maybe it of String data type.
    </span>

In [8]:
carFeatureDf.describe(include = 'all')

car          price   body      mileage         engV engType  \
count         9576    9576.000000   9576  9576.000000  9142.000000    9576   
unique          87            NaN      6          NaN          NaN       4   
top     Volkswagen            NaN  sedan          NaN          NaN  Petrol   
freq           936            NaN   3646          NaN          NaN    4379   
mean           NaN   15633.317316    NaN   138.862364     2.646344     NaN   
std            NaN   24106.523436    NaN    98.629754     5.927699     NaN   
min            NaN       0.000000    NaN     0.000000     0.100000     NaN   
25%            NaN    4999.000000    NaN    70.000000     1.600000     NaN   
50%            NaN    9200.000000    NaN   128.000000     2.000000     NaN   
75%            NaN   16700.000000    NaN   194.000000     2.500000     NaN   
max            NaN  547800.000000    NaN   999.000000    99.990000     NaN   

       registration         year    model  drive  
count          9576  9576.000000     9576   9065  
unique            2          NaN      888      3  
top             yes          NaN  E-Class  front  
freq           9015          NaN      199   5188  
mean            NaN  2006.605994      NaN    NaN  
std             NaN     7.067924      NaN    NaN  
min             NaN  1953.000000      NaN    NaN  
25%             NaN  2004.000000      NaN    NaN  
50%             NaN  2008.000000      NaN    NaN  
75%             NaN  2012.000000      NaN    NaN  
max             NaN  2016.000000      NaN    NaN

- The __Price__ column is having discrepancies. The minimum price is started from zero which is not correct
- Max Prices is ‭32 times greater than 4th. I could have outliers.
- __`Body`__ column is having 6 unique values. 'Sedan' body type is having the highest occurrence. 
- __`Mileage`__ is having outliers. Minimum and maximum mileage are 0 and 999 respectively, which could be wrong data.  
- __`engType`__ is having 4 unique values. __Petrol__ type engine is having maximum occurrence.
- __`registration`__ is having 2 unique values. Most of the value is 'YES' because its occurrence is 9015 times.
-  __`year`__ is having a huge spread. It has a min from 1953 to max 2016.
- __`model`__ is having 888 unique value.
- __`drive`__ is having 3 unique values. __front__ type is having the maximum occurrence.

#### Initial Observation
The dataset has huge discrepancies,outlires and missing values. We have to do data processing to correct the datase.

In [9]:
carFeatureDf.isnull().sum()

car               0
price             0
body              0
mileage           0
engV            434
engType           0
registration      0
year              0
model             0
drive           511
dtype: int64

<span style='color:blue'>
    
__engV__ and __drive__ column is having null values. 

 </span>

<a id="section402"></a>
#### 4.2 Pre Profiling

- Panda Profile helps to ___visualize the dataset interactively.___
- Provide a __360-degree view of dataset__ and ___granular level analysis of each column___.
- In the Below section, an HTML file will generate which gives ___visual interpretation___ about data.

In [10]:
profile = profiling.ProfileReport(carFeatureDf)
profile.to_file(outputfile="Car_feature.html")

C:\ProgramData\Anaconda3\envs\insiad\lib\site-packages\pandas_profiling\describe.py:392: FutureWarning: The join_axes-keyword is deprecated. Use .reindex or .reindex_like on the result to achieve the same functionality.
  variable_stats = pd.concat(ldesc, join_axes=pd.Index([names]), axis=1)


#### Findings

- 1 % of the data is missing.
- 2.8 % of __Price__ values are Zeros. It also has a huge spread. The min values are less than 400 $. The max value is very high. 
- 3.6 % of __Mileage__ are zeros.It also has a huge spread which is 1 to 999.
- 4.5 % of __engV__ has missing values. It is also having a huge spread e.i. 0.1 to 99.9.
- 5.3 % of __drive__ data is missing. 
- 113 rows are __duplicate__

<a id=section403></a>
#### 4.3 Preprocessing

- Removing Data discrepancy   <br/>

#### Delete Duplicate rows

In [11]:
print((carFeatureDf.sort_values(by= carFeatureDf.columns.tolist() ,ascending=False)).tail(10) ) # It will sort the dataset by all coulumns.

print('=========================================================================================')
print('=========================================================================================')
groupedDf=carFeatureDf.groupby(carFeatureDf.columns.tolist(),as_index=False).size()
(groupedDf.sort_values(ascending=False)).head()

        car    price       body  mileage  engV engType registration  year  \
9063  Acura  18500.0  crossover       85   3.7  Petrol          yes  2008   
6638  Acura  17200.0  crossover       82   3.7  Petrol          yes  2008   
4484  Acura  15650.0  crossover      170   3.7  Petrol          yes  2008   
2198  Acura  15000.0      sedan      150   3.5     Gas          yes  2008   
5835  Acura  12900.0      sedan      126   3.5     Gas          yes  2006   
5452  Acura  11111.0  crossover      199   3.5  Petrol          yes  2005   
8082  Acura   8700.0      sedan      145   3.2     Gas          yes  2005   
7390  Acura   8699.0      sedan      145   3.2     Gas          yes  2005   
4803  Acura   8699.0      sedan      144   3.2     Gas          yes  2005   
7194  Acura   8550.0      sedan      145   3.2  Petrol          yes  2005   

     model  drive  
9063   MDX   full  
6638   MDX   full  
4484   MDX   full  
2198    RL   full  
5835    RL   full  
5452   MDX   full  
8082    TL  

car            price     body       mileage  engV  engType  registration  year  model             drive
Toyota         0.0       crossover  0        4.5   Diesel   yes           2016  Land Cruiser 200  full     5
Mercedes-Benz  0.0       crossover  0        3.0   Diesel   yes           2016  GLE-Class         full     4
               199999.0  crossover  0        5.5   Petrol   yes           2016  GLS 63            full     4
               47600.0   van        0        2.2   Diesel   yes           2016  Vito ïàññ.        rear     4
               20400.0   sedan      190      1.8   Gas      yes           2011  E-Class           rear     3
dtype: int64

In above resul we can see many rows are duplicate.

In [12]:
carFeatureDf =carFeatureDf.drop_duplicates() # duplicate data is removed now.
carFeatureDf.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
9571    False
9572    False
9573    False
9574    False
9575    False
Length: 9463, dtype: bool

- Now Duplicate rows are deleated.
- The final dataset size is 9463.

#### Price Transformation
- Replace __Price__ having __0__ with median base on __Car__ and __Model__

In [13]:
columnsToGroup_Sort =[carFeatureDf.car,carFeatureDf.model]
group_df =carFeatureDf[carFeatureDf['price']==0].groupby(by =columnsToGroup_Sort,as_index=False).size()
group_df =group_df.reset_index() 
group_df  # groupin the data having 'Price' column zero by Car and Model. group_df is having unique car and model data. 

car                   model  0
0          Audi                     100  1
1          Audi                      A4  1
2          Audi                      A6  4
3          Audi              A6 Allroad  3
4          Audi                      A8  5
..          ...                     ... ..
135  Volkswagen  T5 (Transporter) ïàññ.  4
136  Volkswagen   T6 (Transporter) ãðóç  1
137  Volkswagen                 Touareg  1
138         ZAZ             1102 Òàâðèÿ  2
139        ËUAZ              969 Âîëûíü  1

[140 rows x 3 columns]

- Replacing  Price with median 

In [14]:
for i, j in group_df.iterrows(): 
   medianPrice=  carFeatureDf[(carFeatureDf.car ==j.car) & (carFeatureDf.model ==j.model)  & carFeatureDf['price'] >0].price.median()
   #  If A combination of car and model is having only one row and price is zero then meadian will come as nans.In this codition I and deleatint that row.
   if( ~ np.isnan(medianPrice)): 
         mask =(carFeatureDf['car'] ==j.car)  &  (carFeatureDf['model'] ==j.model)  & (carFeatureDf['price'] ==0)
         carFeatureDf['price'][mask]=medianPrice
   else:
    # Deleting the rows which done not have any row as per below codition
     mask =(carFeatureDf['car'] ==j.car)  &  (carFeatureDf['model'] ==j.model)  & (carFeatureDf['price'] ==0)
     carFeatureDf.drop(carFeatureDf[mask].index, inplace = True) 

C:\ProgramData\Anaconda3\envs\insiad\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


#### Replacing 'Other' type Body
- 826 rows are having body type as '__Other__' .I grouped the data with multiple columns and find the below pattern for __`Body`__
    - __Body__ type is most likely same for the same type of __Car__,__model__ and __engType__
    - It is not dependent at __year__ and __price__

In [83]:
columnsToGroup_Sort =[carFeatureDf.car,carFeatureDf.model, carFeatureDf.engType]
group_df =carFeatureDf[(carFeatureDf['body']=='other')].groupby(by =columnsToGroup_Sort,as_index=False).size()
group_df =group_df.reset_index() 
group_df.sort_values([0,'car','model'], axis=0, ascending=False) .head(50)
group_df

car   model engType  0
0    Aston Martin     DB9  Petrol  1
1            Audi     100     Gas  1
2            Audi      80  Petrol  1
3            Audi      A4  Diesel  1
4            Audi      A4  Petrol  1
..            ...     ...     ... ..
426           ZAZ  110557     Gas  3
427           ZAZ  110557  Petrol  3
428           ZAZ     965  Petrol  1
429           ZAZ     968  Petrol  1
430          ËUAZ    969Ì  Petrol  1

[431 rows x 4 columns]

- Replacing 'Other' type Body

In [117]:
    for i, j in group_df.iterrows(): 
       body_mode=  carFeatureDf[ ~(carFeatureDf['body']=='other') & (carFeatureDf.car ==j.car) & (carFeatureDf.model ==j.model)  & (carFeatureDf['engType'] ==j.engType)].body.mode()
       if(len(body_mode)>0):
              mask =((carFeatureDf['body']=='other') & (carFeatureDf.car ==j.car) & (carFeatureDf.model ==j.model))
              carFeatureDf['body'][mask]=body_mode[0]
       else: 
        mask =((carFeatureDf['body']=='other') & (carFeatureDf.car ==j.car) & (carFeatureDf.model ==j.model))
        carFeatureDf.drop(carFeatureDf[mask].index, inplace = True) 

C:\ProgramData\Anaconda3\envs\insiad\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


In [118]:
carFeatureDf[carFeatureDf['body']=='other']

Empty DataFrame
Columns: [car, price, body, mileage, engV, engType, registration, year, model, drive]
Index: []